### DISCUSSION TITLE
mcp23x17 help needed

### DISCUSSION THREAD

**Discussion by aeropagz** [UncleRus/esp-idf-lib#312] [Q&A]
Hello,

I am using your mcp23x17 lib to read 12 magnetic sensors at once. When I set the mode of all 16 gpio to input mode, the 8th bit is not set to input. Is there bug ?

The port expander: MCP23S17-E/SS
This is my code so far:

```C
#include <stdio.h>
#include <freertos/FreeRTOS.h>
#include <freertos/task.h>
#include <mcp23x17.h>
#include <driver/gpio.h>
#include "esp_log.h"
#include <string.h>

#include "goal_slider.h"

#define PIN_NUM_MOSI 13
#define PIN_NUM_MISO 12
#define PIN_NUM_CLK 14
#define PIN_NUM_CS 5

#define HOST HSPI_HOST

static const char *TAG = "GOAL SLIDER";

static mcp23x17_t dev = {0};

void goal_slider_read(void *pvParameters)
{
    uint16_t val;
    // Configure SPI bus
    spi_bus_config_t cfg = {
        .mosi_io_num = PIN_NUM_MOSI,
        .miso_io_num = PIN_NUM_MISO,
        .sclk_io_num = PIN_NUM_CLK,
        .quadwp_io_num = -1,
        .quadhd_io_num = -1,
        .max_transfer_sz = 0,
        .flags = 0
    };

    ESP_ERROR_CHECK(spi_bus_initialize(HOST, &cfg, 1));
    ESP_ERROR_CHECK(mcp23x17_init_desc_spi(&dev, HOST, MCP23X17_MAX_SPI_FREQ, 0x20, PIN_NUM_CS));

    ESP_ERROR_CHECK(mcp23x17_set_int_out_mode(&dev, MCP23X17_ACTIVE_LOW));
    
    // weird stuff happens here
    ESP_ERROR_CHECK(mcp23x17_port_set_mode(&dev, 0xFFFF));
    ESP_ERROR_CHECK(mcp23x17_port_get_mode(&dev, &val));
    ESP_LOGI(TAG, "Mode %x\n", val); // returns 0xff7f

    while (1)
    {
        esp_err_t ret = mcp23x17_port_read(&dev, &val);
        assert(ret == ESP_OK);
        ESP_LOGI(TAG, "Value read: %x\n", val);
        vTaskDelay(1000 / portTICK_PERIOD_MS);
    }
}
```

Could you give me some advice or help. I would appreciate it :)

**Comment by UncleRus** [comment]
Hi!

First try another MCP23S17.

**Reply by aeropagz** [reply]
I tried with another MCP23S17. Now I get other errors. For Example:

Sensor Active:
`0000 0000 0000 0011`

Input Read:
`1000 0000 0000 0001`

Furthermore the 4th most significant bits/inputs are not connected to any sensors (can not change).

Other Example:

Sensor Active:
`0000 0001 1000 0000`

Input Read:
`0000 0000 0100 0000`

BUT sometimes it is right as well: 


Sensor Active:
`0000 0000 0000 0110`

Input Read:
`0000 0000 0000 0110`

Feels like the order of inputs is messed up or not working. Quite strange, any idea what is wrong?

**Comment by trombik** [comment]
> Furthermore the 4th most significant bits/inputs are not connected to any sensors (can not change).

IIRC, floating means it can be HIGH or LOW at any time. if you don't use them, ignore or ground them. i would use oscilloscope or logic analyzer to see what is actually happening.

**Reply by aeropagz** [reply]
they are grounded. It can't be a hardware problem. It works perfectly with Arduino Libary. There must be some misconfiguration in my code.

**Comment by aeropagz** [comment]
```
    ESP_ERROR_CHECK(mcp23x17_port_set_mode(&dev, 0xFFFF));
    ESP_ERROR_CHECK(mcp23x17_port_get_mode(&dev, &val));
    
    ESP_LOGI(TAG, "Mode %d\n", val);
```

I tried this and val returns `FF7F`. So somehow one bit is not set into input mode. Same happens one the other board.
